packages/web-components/fast-ssr/src/request-storage.ts

```diff
@@ -156,18 +185,54 @@ export const RequestStorageManager = Object.freeze({
 
     /**
      * Installs a DOM shim that ensures that window, document,
-     * and other globals are scoped per-request.
+     * and other globals are scoped per-request. Calling this function
+     * will have no effect if the shim has already been installed.
+     *
+     * @throws TypeError when properties cannot be defined on the globalThis.
      */
     installDOMShim(): void {
         for (const key of perRequestGlobals) {
-            const original = (globalThis as any)[key];
-            Reflect.defineProperty(globalThis, key, {
-                get() {
-                    // Return original global variable if currently not in the storage scope
-                    const store = asyncLocalStorage.getStore() as Map<string, any>;
-                    return store ? store.get("window")[key] : original;
-                },
-            });
+            if (!shimIsInstalledFor(key)) {
+                const preShimValue = (globalThis as any)[key];
+                preShimGlobals.set(key, (globalThis as any)[key]);
+                const preShimDescriptor = Object.getOwnPropertyDescriptor(
+                    globalThis,
+                    key
+                );
+
+                // This will throw if the globalThis already contains a value for the key that is not configurable. Do this work
+                // prior to caching value and descriptor so if it does throw, the caches aren't polluted
+                Object.defineProperty(globalThis, key, {
+                    get: perRequestGetters[key],
+                    enumerable: true,
+                    configurable: true,
+                });
+
+                if (preShimDescriptor) {
+                    preShimDescriptors.set(key, preShimDescriptor);
+                }
+
+                preShimGlobals.set(key, preShimValue);
+            }
+        }
+    },
+
+    /**
+     * Uninstalls the DOM shim installed by {@link RequestStorageManager.installDOMShim}.
+     * Calling this function will have no effect if there is no shim installed.
+     */
+    uninstallDOMShim(): void {
+        for (const key of perRequestGlobals) {
+            if (shimIsInstalledFor(key)) {
+                if (preShimDescriptors.has(key)) {
+                    Object.defineProperty(globalThis, key, preShimDescriptors.get(key)!);
+                    preShimDescriptors.delete(key);
+                } else {
+                    delete (globalThis as any)[key];
```

Comment:
Will this line ever be hit? And should we care to delete this global variable? When `installDOMShim` throws on `defineProperty`, `perRequestGetters[key]` is not set on the object right? If so, `preShimDescriptors` should always have the key cached with descriptor.